lib/units/storage/amazons3.js

```diff
@@ -0,0 +1,197 @@
+var http = require('http')
+var util = require('util')
+var path = require('path')
+
+var express = require('express')
+var validator = require('express-validator')
+var bodyParser = require('body-parser')
+var formidable = require('formidable')
+var Promise = require('bluebird')
+var uuid = require('node-uuid')
+var fs = require('fs')
+
+var lifecycle = require('../../util/lifecycle')
+var logger = require('../../util/logger')
+var requtil = require('../../util/requtil')
+
+var AWS = require('aws-sdk')
+
+module.exports = function(options) {
+  var log = logger.createLogger('storage:s3')
+  , app = express()
+  , server = http.createServer(app)
+  , credentials = new AWS.SharedIniFileCredentials({
+      profile: options.profile
+    })
+  , bucket = options.Bucket
+  , s3
+
+  // TODO: remove later
+  log.debug(options)
+
+  AWS.config.credentials = credentials;
+  s3 = new AWS.S3(options)
+
+  app.set('strict routing', true)
+  app.set('case sensitive routing', true)
+  app.set('trust proxy', true)
+
+  app.use(bodyParser.json())
+  app.use(validator())
+
+  function putObject(plugin, uriorpath, name) {
```

Comment:
I would maybe change putObject into something like this:

```js
function putObject(plugin, file) {
  return new Promise(function(resolve, reject) {
    var id = uuid.v4()
    s3.putObject({
      Key: id
    , Body: fs.createReadStream(file.path)
    , Bucket: options.bucket
    , Metadata: {
        plugin: plugin
      , name: file.name
      }
    }, function(err, data) {
      if (err) {
        log.error('Unable to store "%s"', file.path, err)
        return reject(err)
      }

      log.info('Stored "%s" as "%s/%s"', file.path, options.bucket, id)
      resolve(id)
    })
  })
}
```

And then use it later like this:

```js
return putObject(plugin, file)
  .then(function(id) {
    return {
      field: field
    , id: id
    , name: file.name
    }
  })
```